[3주차] 예매 API 구현 및 동시성 문제해결

kimbro97 opened the pull request:
### [3주차] 예매 API 구현 및 동시성 문제해결

<br/>  

### 작업 내용  
-  예매 API 기능 개발
- 
-

### 발생했던 문제와 해결 과정을 남겨 주세요. 
- 비관적 락 테스트 오류
    -  현재 예매 기능은 Reservation과 ReservedSeat 엔티티에 예매된 좌석 정보를 저장하도록 설계되어 있습니다.
    -  예매 로직에서 reservedSeatRepository.findByScreeningAndSeats 메소드를 호출하여 해당 좌석이 이미 예매되었는지 확인합니다.
    -  이를 통해 이미 예매된 좌석에 대해서는 예매가 불가능하도록 처리하고 있습니다.
    - 동시성 문제를 방지하기 위해 findByScreeningAndSeats 메소드에 비관적 락(FOR UPDATE)을 설정했습니다.
    - 테스트 코드를 실행했을 때, 쿼리에 FOR UPDATE 문이 포함되어 정상적으로 실행되는 것을 확인했습니다.
    - 하지만 동시성 문제를 제어하지 못하는 것으로 보입니다.
    - 어떤게 잘못된건지 잘모르겠습니다.
-
-

### 이번 주차에서 고민되었던 지점이나, 어려웠던 점을 알려 주세요.
-  
-  
-    

### 리뷰 포인트
> 리뷰어가 특히 의견을 주었으면 하는 부분이 있다면 작성해 주세요.<br/>
-  예매 로직에서 검증(Validation) 로직이 많아졌습니다. 이를 관리하기 위해 ReservationValidator라는 별도 객체를 생성하여 검증 로직을 분리하였습니다. 이 접근 방식이 적절한지, 혹은 더 나은 방식이 있을지 궁금합니다!
-
-
-  

### 기타 질문
> 추가로 질문하고 싶은 내용이 있다면 남겨주세요.<br/>
> ex) 테스트 환경에서 동시성 테스트를 수행하였고, 모든 케이스를 통과했습니다. 추가할 테스트 시나리오가 있을까요?
-
-

kimbro97 committed:
chore: 스키마 및 더미데이터 수정

kimbro97 committed:
chore: 빌드 오류 수정

kimbro97 committed:
refactor: 불필요한 테스트코드 삭제

kimbro97 committed:
refactor: 패키지 구분 및 Reservation entity 추가

kimbro97 committed:
test: 5개 이상 예매시 예외 발생 테스트 코드 작성

kimbro97 committed:
refactor: Seats 일급컬렉션 도입 -> 캡슐화 및 책임과 역할 분리

kimbro97 committed:
refactor: Seats 불변객체 작업

kimbro97 committed:
feat: 연속적인 좌석 유효성 검사 작업

kimbro97 committed:
chore: 데이터 및 스키마 추가

kimbro97 committed:
chore: reservation http 추가

kimbro97 committed:
feat: 예약 API 구현

kimbro97 committed:
refactor: User -> Member로 변경

kimbro97 committed:
test: ReservationService service 로직 테스트코드 작성

kimbro97 committed:
refactor: 테스트코드 기반으로 서비스로직 리팩터링

kimbro97 committed:
feat: pessimistic lock 구현 중

kimbro97 committed:
chore: README 파일 수정

pyuseon commented:
안녕하세요 형재님 이번 주차고 고생 많으셨습니다. 
리뷰 진행하도록 하겠습니다.

pyuseon commented on api/src/main/java/com/example/reservation/request/ReservationRequest.java:
```diff
@@ -0,0 +1,39 @@
+        if (seatIds.isEmpty() || seatIds.size() > 5) {
+            throw new IllegalArgumentException("예매할 좌석을 1개 이상 5개 이하로 선택해주세요.");
+        }
+    }
```
`ReservationRequest`는 클라이언트 요청 데이터를 단순히 담는 DTO(Data Transfer Object) 역할은 하는 객체 입니다. 
그러나 현재 로직에서는 데이터를 검증하고 변환하는 책임까지 포함되어 있어 **단일 책임 원칙(Single Responsibility Principle)** 에 위반되는 부분입니다! 
검증 로직이 DTO 내부에 포함될 경우, 검증 로직을 수정하거나 유지보수할 때 `ReservationValidator`와 `ReservationRequest` 두 곳을 모두 확인해야 하는 불편함이 생깁니다.
이를 방지하고 유지보수성을 높이기 위해, 검증 로직은 `ReservationValidator`로 이동시켜 DTO는 데이터 전달과 변환 역할만 담당하도록 수정해 주세요!

pyuseon commented on application/src/main/java/com/example/reservation/validator/ReservationValidate.java:
```diff
@@ -0,0 +1,79 @@
+                .map(seat -> new ReservedSeat(reservation, seat))
+                .toList();
+        reservation.addReservedSeat(reservedSeats);
+
```
검증 로직 분리에 대해 질문 주셨는데요, 현재 구조에서는 검증 로직과 예약 객체 생성 로직이 혼합되어 있어 책임이 명확히 구분되지 않고 있습니다.
검증 로직은 ReservationValidator에서 한 곳에서만 관리하도록 하고, 예약 객체 생성 부분은 Service로 이동시키는 것이 더 책임을 명확히 구분하는 설계가 될 것입니다.

```
java

List<ReservedSeat> existingReservations = reservedSeatRepository.findByScreeningAndSeats(screening, seats.getSeats());
if (!existingReservations.isEmpty()) {
    throw new IllegalArgumentException("이미 예매된 좌석입니다.");
}
```
이 부분은 현재 락이 걸리는 구간인데요, Service 단에서 실행되도록 이동하여 검증과 예약 로직 전체를 락으로 보호받도록 설계하시는 것을 추천드립니다.
이렇게 하면 동시성 문제를 방지하는데 도움이 될거에요!

pyuseon commented on application/src/main/java/com/example/reservation/validator/ReservationValidate.java:
```diff
@@ -0,0 +1,79 @@
+        return reservedSeatRepository.findAllByMemberId(member, screening);
+    }
+
+    private Member getMember(Long memberId) {
+        return memberRepository.findById(memberId).orElseThrow(() -> new IllegalArgumentException("회원 정보가 없습니다."));
+    }
```
클라이언트에서 받은 회원 정보를 한번 더 확인해 주고 계시네요! 
시스템 간 통신에서 데이터 신뢰성이 중요하기 때문에, 이와 같은 재검증은 실무에서도 안정성을 높이는 좋은 방법입니다!

pyuseon commented on api/src/main/java/com/example/reservation/ReservationController.java:
```diff
@@ -0,0 +1,20 @@
+
+    private final ReservationService reservationService;
+
+    @PostMapping("/reservation")
```
api 에는 버전 정보를 같이 명시해 주시는 것이 좋습니다. v1/reservation 과 같이요!
실무에서는 버전 관리를 위해 api에 명시해 주는 경우가 많습니다.

pyuseon commented on domain/src/main/java/com/example/repository/reservation/ReservedSeatRepositoryImpl.java:
```diff
@@ -0,0 +1,39 @@
+                .setLockMode(LockModeType.PESSIMISTIC_WRITE)
+                .fetch();
+    }
+
```
현재 이 부분에 비관적 락을 걸어 주신 것으로 보이는데요 
`ReservationValidate.validate()` 내부에서 findByScreeningAndSeats를 호출하여 락을 설정했더라도, 락이 풀리는 시점과 데이터 수정 시점이 일치하지 않을 수 있습니다.
예약 작업의 가장 초기에 락을 설정하여 검증 로직과 예약 생성 로직까지 동일한 트랜잭션 내에서 실행되도록 수정해 보시는 것은 어떨까요?

pyuseon commented on domain/src/main/java/com/example/entity/reservation/Reservation.java:
```diff
@@ -0,0 +1,57 @@
+        if (!seats.isContinuousSeat()) {
+            throw new IllegalArgumentException("좌석 예매는 연속적인 좌석만 예매 가능합니다.");
+        }
+    }
```
Entity 클래스에 검증 로직에 대한 책임을 두면 유지 보수가 어려워 질수 있어요! 이부분도 validator에서 관리해 주시는 것이 좋습니다!

pyuseon commented on application/src/test/java/com/example/reservation/ReservationServiceTest.java:
Validation 로직에 있는 내용들을 테스트 코드에 잘 반영해 주셨네요!
덕분에 추후 코드가 수정되더라도 validation 로직의 검증을 꼼꼼히 확인할 수 있을 것 같습니다. 😄

pyuseon commented:
# 좋았던점 
* 데이터 검증을 꼼꼼하게 작성해 주셨습니다. 클라이언트에서 받아오는 정보도 한번 더 확인하도록 설계해 주셔서 견고한 서버가 될 것 같습니다. 
* 테스트 코드를 여러 케이스별로 작성해 주셨는데요, 이를 통해 Validation 검증 로직에 오류가 없는지 충분히 확인할 수 있을 것 같네요. 훌륭합니다! 😊

# 아쉬운점 
* 시나리오 요구사항 중 '예약이 완료되면 FCM을 통한 App Push 발송' 부분이 현재 반영되지 않은 것 같습니다. 이 부분을 추가로 구현하시면 좋을 것 같습니다.
* 검증을 위한 코드가 여러 클래스에 침투해 있어, 객체의 책임이 분리되지 않는 문제가 있습니다. 이 부분에 대해서는 코드 리뷰에 구체적으로 작성해 두었으니, 리팩토링을 통해 개선해 보세요! 

# 리뷰 포인트 
* 현재 진행중인 비관적락은 리뷰 내용을 참고해 수정해 보세요! [jpa-pessimistic-locking](https://www.baeldung.com/jpa-pessimistic-locking) 링크도 한번 참고해 보시는 것이 좋아요. 
* 추가적으로 낙관적락과 분산락도 적용해 보시는 것을 추천 드립니다. 


# 추가 질문 사항 
* 코드 리뷰에 관련 내용을 작성해 두었습니다. 

추가 질문이 있으시면 이 pr에 멘션으로 코멘트 달아주세요! 
연휴에도 과제하시느라 고생 많으셨습니다! 새해 복 많이 받으세요~ 🙇‍♀️

kimbro97 committed:
refactor: API endpoint 버전 명시

kimbro97 committed:
refactor: DTO 검증 로직 이동

kimbro97 committed:
refactor: pessimistic lock 구현
- 기존에는 ReservedSeatRepository.findByScreeningAndSeats 메서드에 비관적 락을 적용했었습니다
- 그러나 비관적 락은 데이터 조회 시, 해당 데이터가 없을 경우 락이 걸리지 않는다는 문제를 확인했습니다.
- 이를 해결하기 위해, SeatRepository.findAllByIdsWithLock 메서드에서 좌석을 조회할 때 비관적 락을 적용하도록 변경했습니다.

kimbro97 committed:
refactor: 검증 로직과 객체 생성 로직의 책임 분리 작업

kimbro97 committed:
feat: MessageService 인터페이스 및 LogMessageService 구현

- MessageService 인터페이스 생성
  - 메시지 발송 로직의 유연성과 확장성을 고려한 설계
  - 실제 FCM 연동 없이 로그 출력을 대체 구현할 수 있도록 구조 설계

- LogMessageService 클래스 구현
  - 메시지 발송 시 로그 출력 기능 제공
  - MessageService 인터페이스를 구현하여 주입 가능한 방식으로 설계

kimbro97 committed:
feat: optimistic lock 구현

- redis Lua script를 활용해 낙관적 락을 구현했습니다.
- 기존에 비관락을 통과한 테스트코드를 비관락 없이 통과했습니다.

kimbro97 committed:
refactor: entity 검증 로직 Vaildate 객체로 이동

kimbro97 commented:
안녕하세요, 코치님!

락 구현이 처음이라 시간이 많이 소요되었지만, 명절임에도 자세한 리뷰를 남겨주셔서 진심으로 감사드립니다.

- 유효성 검사 관련 로직을 ReservationValidate 객체에서 통합적으로 관리하도록 수정했습니다.
- 비관적 락을 적용할 때, 예매된 좌석인지 확인하는 쿼리에 락을 걸었는데, 데이터를 조회할 수 없을 경우 락이 걸리지 않는다는 점을 뒤늦게 인지했습니다.
- 이를 해결하기 위해 선택한 좌석을 조회하는 과정에서 비관적 락을 적용하여 동시성 문제를 방지했습니다.
- FCM 발송 서비스는 MessageService 인터페이스와 LogMessageService 구현체를 분리하여 객체 간 결합도를 최소화했습니다.
- 이를 통해, 향후 FCM 발송 서비스를 실제로 적용하더라도 클라이언트 코드의 변경 없이 사용할 수 있도록 설계했습니다.
- 추가로, Redis Lua 스크립트를 활용하여 낙관적 락을 구현해보았습니다.

현재까지 낙관적 락까지 개발을 마쳤으며, 남은 부분은 4주차 과제와 함께 꾸준히 진행해 나가겠습니다.
머지 부탁드리며, 새해 복 많이 받으세요! 😊🎉
